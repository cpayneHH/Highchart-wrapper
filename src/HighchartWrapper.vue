<!-- src/HighchartWrapper.vue -->
<template>
  <div class="highchart-wrapper">
    <!-- Always show the title title -->
    <h2 v-if="props?.title">{{ props.title }}</h2>

    <template v-if="isLoading">
      <div class="p-4 text-center text-gray-400">Loading data, please wait...</div>
    </template>

    <template v-else-if="isError">
      <div class="p-4 text-center text-red-500">
        Failed to load data.
        <button @click="reloadData" class="mt-2 btn btn-error btn-sm">Retry</button>
      </div>
    </template>

    <template v-else-if="!shouldShowFallback || hasData">
      <!-- If a non-empty layout was provided, use the grid-based layout -->
      <GridLayout v-if="props.layout?.length" :layout="props.layout" :items="dataItems" />

      <!-- Otherwise, render items one-by-one in the default BlockRenderer -->
      <BlockRenderer v-else :items="dataItems" />
    </template>

    <template v-else>
      <div class="fallback-message p-4 text-center text-gray-500">
        Please provide an items object or API endpoint to obtain data from.
      </div>
    </template>
  </div>
</template>

<script setup lang="ts">
import { computed, defineProps, onMounted, ref } from 'vue';
import type { BlockItem, WrapperProps } from './types';
import { fetchItems } from './services/api';

import GridLayout from './components/GridLayout.vue';
import BlockRenderer from './components/BlockRenderer.vue';

const isLoading = ref(false);
const isError = ref(false);

// 1) Grab props as defined in your types.ts
const props = defineProps<WrapperProps>();

// 2) Create a local, reactive copy of items
//    (start with whatever the consumer passed in)
const dataItems = ref<BlockItem[]>(props.items);

// 3) If no items were passed, but an endpoint exists, fetch on mount
onMounted(async () => {
  if ((!props.items || props.items.length === 0) && props.endpoint) {
    isLoading.value = true;
    isError.value = false;
    try {
      const responses = await fetchItems(props.endpoint);
      // flat‐map all of their payload arrays into dataItems
      dataItems.value = responses.flatMap((r) => r.payload);
    } catch (err) {
      console.error('HighchartWrapper: failed to fetch items from', props.endpoint, err);
      isError.value = true;
    } finally {
      isLoading.value = false;
    }
  }
});

const hasData = computed(() => dataItems.value?.length > 0);
const shouldShowFallback = computed(() => {
  return (!props.items || props.items.length === 0) && !props.endpoint;
});

async function reloadData() {
  if (!props.endpoint) return;

  isLoading.value = true;
  isError.value = false;

  try {
    const responses = await fetchItems(props.endpoint);
    dataItems.value = responses.flatMap((r) => r.payload);
  } catch (err) {
    console.error('HighchartWrapper: failed to refetch items from', props.endpoint, err);
    isError.value = true;
  } finally {
    isLoading.value = false;
  }
}
</script>

<style>
/* Component styles */
.highcharts-container {
  .highcharts-root {
    padding: 1rem;
  }

  .highcharts-credits {
    display: none !important;
  }
}

.dashboard-tile-height {
  height: 400px;
}

.dashboard-tile-shadow {
  box-shadow: 0 0 3px rgba(0, 0, 0, 0.2);
}
</style>
