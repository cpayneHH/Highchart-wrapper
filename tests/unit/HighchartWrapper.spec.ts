import {afterEach, beforeAll, describe, expect, it, vi} from 'vitest';
import {render, screen, waitFor, within} from '@testing-library/vue';
import '@testing-library/jest-dom';
import HighchartWrapper from '@components/HighchartWrapper.vue';
import {server} from '@mocks/server';
import Highcharts from '@mocks/highcharts';
import APIResponseSingle from '@mocks/api.response.block.json';
import APIResponseMultiple from '@mocks/api.response.blocks.json';
import {BlockItem} from '@types';
import {http, HttpResponse} from 'msw';

describe('HighchartWrapper Component Tests', () => {
    const requests: Request[] = [];

    // Mock Highcharts and its modules
    vi.mock('highcharts', async () => await import('../mocks/highcharts'));
    vi.mock('highcharts/modules/pattern-fill');
    vi.mock('highcharts/modules/exporting');
    vi.mock('highcharts/modules/accessibility');
    vi.mock('highcharts/modules/sankey');
    vi.mock('highcharts/modules/funnel');

    beforeAll(() => {
        server.events.on('request:match', (req) => requests.push(req.request));
    });

    afterEach(() => {
        // Clear tracked requests and mock calls
        requests.length = 0;
        Highcharts.chart.mockClear();
    });

    // Snapshot Test
    it('T001 - Snapshot: Full dashboard layout', async () => {
        const {container} = render(HighchartWrapper, {
            props: {
                title: 'Dashboard Snapshot',
                endpoint: '/api/v1/blocks/retrieve',
                layout: [[2, 1], [1, 1, 1], [1, 2], [2, 1], [3]],
            },
        });

        // Wait for all chart blocks to initialize
        await waitFor(() => {
            expect(Highcharts.chart).toHaveBeenCalledTimes(
                APIResponseMultiple.payload.filter(b => b.type === 'chart').length
            );
        });

        // Compare rendered output to snapshot
        expect(container).toMatchSnapshot();
    });

    // Prop Defaults
    it('T002 - Default Props: handles missing all props', () => {
        render(HighchartWrapper);

        // No level-2 heading when title omitted
        expect(screen.queryByRole('heading', {level: 2})).not.toBeInTheDocument();

        // Fallback message appears
        expect(
            screen.getByText(
                'Please provide an items array of objects or API endpoint to obtain data from.'
            )
        ).toBeInTheDocument();
    });

    // Partial Props
    it('T003 - Partial Props: title undefined, renders provided item', async () => {
        render(HighchartWrapper, {
            props: {
                items: [APIResponseSingle.payload[0]] as BlockItem[],
                layout: null,
            }
        });

        // No heading rendered
        expect(screen.queryByRole('heading', {level: 2})).not.toBeInTheDocument();

        // Chart stub should appear
        const chart = await screen.findByTestId('chart');
        expect(chart).toBeInTheDocument();
        expect(Highcharts.chart).toHaveBeenCalled();
    });

    it('T004 - Fetch via Endpoint: layout undefined', async () => {
        render(HighchartWrapper, {props: {endpoint: '/api/v1/blocks/retrieve'}});

        // Wait for all charts
        await waitFor(() => {
            expect(Highcharts.chart).toHaveBeenCalledTimes(
                APIResponseMultiple.payload.filter(b => b.type === 'chart').length
            );
        });
    });

    // Title Rendering
    it('T005 - Title Rendering: no h2 when title is null', () => {
        render(HighchartWrapper, {
            props: {title: null, endpoint: null, items: null, layout: null},
        });

        expect(
            screen.queryByRole('heading', {level: 2})
        ).not.toBeInTheDocument();
    });

    it('T006 - Title Rendering: renders h2 when title provided', () => {
        render(HighchartWrapper, {props: {title: 'My Title', items: [], layout: []}});

        expect(
            screen.getByRole('heading', {level: 2, name: 'My Title'})
        ).toBeInTheDocument();
    });

    // Fallback UI
    it('T007 - Fallback UI: shows message when no items or endpoint', () => {
        const {container} = render(HighchartWrapper, {
            props: {title: null, endpoint: null, items: null, layout: null},
        });

        const fallback = container.querySelector('.fallback-message');
        within(<HTMLElement>fallback).getByText(
            'Please provide an items array of objects or API endpoint to obtain data from.'
        );
    });

    // Full Endpoint-driven Render
    it('T008 - Full Render: displays grid and blocks from endpoint', async () => {
        render(HighchartWrapper, {
            props: {
                endpoint: '/api/v1/blocks/retrieve',
                layout: [[2, 1], [1, 1, 1], [1, 2], [2, 1], [3]],
            }
        });

        await waitFor(() => {
            // Verify network request
            expect(requests.length).toBeGreaterThan(0);
            const req = requests.find(r => r.url.includes('/api/v1/blocks/retrieve'));
            expect(req).toBeDefined();
        });

        // Inspect grid layout
        const grid = screen.getByTestId('grid-layout');

        // Rows count
        const rows = await within(grid).findAllByTestId('row');
        expect(rows).toHaveLength(5);

        // Columns count
        const cols = await within(grid).findAllByTestId('col');
        expect(cols).toHaveLength(10);

        // Title and callout counts
        const titles = await within(grid).findAllByTestId('title');
        expect(titles).toHaveLength(2);
        const callouts = await within(grid).findAllByTestId('callout');
        expect(callouts).toHaveLength(2);

        // Charts rendered
        expect(Highcharts.chart).toHaveBeenCalled();
        const charts = await within(grid).findAllByTestId('chart');
        expect(charts).toHaveLength(6);
    });

    // Single-item Render
    it('T009 - Single Item: renders one chart via items prop', () => {
        const {container} = render(HighchartWrapper, {
            props: {title: 'Single Chart', items: [APIResponseSingle.payload[0]] as BlockItem[]}
        });

        // Verify heading text
        within(container.querySelector('h2')!).getByText('Single Chart');

        // Container role
        expect(container.querySelector('[role="region"]')).toBeDefined();
    });

    // Error Handling
    it('T010 - Error & Retry: shows error and retries fetch', async () => {
        // Mock server to throw
        server.use(
            http.get('/api/v1/blocks/retrieve', ({request, params, cookies}) => HttpResponse.error())
        );

        render(HighchartWrapper, {props: {endpoint: '/api/v1/blocks/retrieve'}});

        // Error state appears
        const errMsg = await screen.findByText(/Failed to load data/i);
        expect(errMsg).toBeInTheDocument();

        // Mock success on retry
        server.use(
            http.get('/api/v1/blocks/retrieve', () => HttpResponse.json(APIResponseSingle))
        );

        // Click retry
        const retryBtn = screen.getByRole('button', {name: /Retry/i});
        await retryBtn.click();

        // Ensure error cleared and chart called
        await waitFor(() => {
            expect(screen.queryByText(/Failed to load data/i)).not.toBeInTheDocument();
            expect(Highcharts.chart).toHaveBeenCalled();
        });
    });

    // Layout Overflow/Underflow
    it('T011 - Layout Overflow: ignores extra items beyond layout slots', async () => {
        render(HighchartWrapper, {
            props: {
                items: [
                    APIResponseMultiple.payload[0],
                    APIResponseMultiple.payload[0],
                    APIResponseMultiple.payload[0]
                ] as BlockItem[],
                layout: [[1, 1], [1]]
            }
        });

        const grid = screen.getByTestId('grid-layout');
        const cols = await within(grid).findAllByTestId('col');
        expect(cols).toHaveLength(3);
        expect(Highcharts.chart).toHaveBeenCalledTimes(3);
    });

    it('T012 - Layout Underflow: only renders available items', async () => {
        render(HighchartWrapper, {
            props: {
                items: [APIResponseSingle.payload[0]] as BlockItem[],
                layout: [[1, 1, 1]]
            }
        });

        const grid = screen.getByTestId('grid-layout');
        const charts = await within(grid).findAllByTestId('chart');
        expect(charts).toHaveLength(1);
        expect(Highcharts.chart).toHaveBeenCalledTimes(1);
    });

    it('T013 - Option Sanitization: strips axes and pointStart for non-axis charts', async () => {
        const pieBlock: BlockItem = {
            id: 11,
            type: 'chart',
            title: 'Pie Chart',
            options: {
                chart: {type: 'pie'},
                xAxis: {title: {text: 'REMOVE ME'}},
                yAxis: {title: {text: 'REMOVE ME TOO'}},
                plotOptions: {series: {pointStart: 5}},
                series: [{type: 'pie', name: 'A', data: [1, 2, 3]}],
            },
        };

        Highcharts.chart.mockClear();
        render(HighchartWrapper, {
            props: {items: [pieBlock], layout: null}
        });

        // Wait for chart initialization
        await waitFor(() => expect(Highcharts.chart).toHaveBeenCalled());

        const opts = Highcharts.chart.mock.calls[0][1];
        expect(opts.xAxis).toBeUndefined();
        expect(opts.yAxis).toBeUndefined();
        expect(opts.plotOptions?.series?.pointStart).toBeUndefined();
    });

});
