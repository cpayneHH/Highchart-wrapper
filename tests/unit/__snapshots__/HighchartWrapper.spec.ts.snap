// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HighchartWrapper Component Tests > T001 - Snapshot: Full dashboard layout 1`] = `
<div>
  <div
    class="highchart-wrapper container"
  >
    <!-- Always show the title title -->
    <h2
      aria-roledescription="Dashboard Snapshot"
      role="heading"
    >
      Dashboard Snapshot
    </h2>
    <!-- isLoading data -->
    
    <!-- Show gridlayout or singular element -->
    <!-- If a non-empty layout was provided, use the grid-based layout -->
    <div
      class="grid-layout"
      data-testid="grid-layout"
    >
      
      <div
        class="row"
        data-testid="row"
      >
        
        <div
          class="my-2 col-lg-8"
          data-testid="col"
        >
          <div
            data-testid="block-renderer"
          >
            
            <div
              aria-label="Symptom Severity Change"
              class="highcharts-block dashboard-tile-shadow"
              data-testid="chart-region"
              role="region"
            >
              <div
                data-testid="chart"
              />
            </div>
            
          </div>
        </div>
        <div
          class="my-2 col-lg-4"
          data-testid="col"
        >
          <div
            data-testid="block-renderer"
          >
            
            <div
              aria-label="Addiction Medicine Program"
              class="callout-block w-100 d-flex align-items-center justify-content-center dashboard-tile-height"
              data-testid="callout"
              role="region"
            >
              <div
                class="px-4"
              >
                <div
                  class="text-mulberry-dark font-size-md pb-2"
                >
                  Addiction Medicine Program
                </div>
                <div
                  class="text-grey-darkest font-size-xs"
                >
                  Combines 24/7 medical care with a highly structured program using proven therapies for behavioural change over six weeks
                </div>
              </div>
            </div>
            
          </div>
        </div>
        
      </div>
      <div
        class="row"
        data-testid="row"
      >
        
        <div
          class="my-2 col-lg-4"
          data-testid="col"
        >
          <div
            data-testid="block-renderer"
          >
            
            <div
              aria-roledescription="-35%"
              class="title-block w-100 d-flex align-items-center justify-content-center dashboard-tile-height dashboard-tile-shadow"
              data-testid="title"
              role="region"
            >
              <div
                class="px-4"
              >
                <h1
                  class="display-4 fw-bold header text-center"
                >
                  -35%
                </h1>
                <div
                  class="text small fw-light text-center"
                >
                  Average Score Improvement
                </div>
              </div>
            </div>
            
          </div>
        </div>
        <div
          class="my-2 col-lg-4"
          data-testid="col"
        >
          <div
            data-testid="block-renderer"
          >
            
            <div
              aria-roledescription="8"
              class="title-block w-100 d-flex align-items-center justify-content-center dashboard-tile-height dashboard-tile-shadow"
              data-testid="title"
              role="region"
            >
              <div
                class="px-4"
              >
                <h1
                  class="display-4 fw-bold header text-center"
                >
                  8
                </h1>
                <div
                  class="text small fw-light text-center"
                >
                  # of Patient that Complete Assessment
                </div>
              </div>
            </div>
            
          </div>
        </div>
        <div
          class="my-2 col-lg-4"
          data-testid="col"
        >
          <div
            data-testid="block-renderer"
          >
            
            <div
              aria-label="Assessment Scores by Assessment Number"
              class="highcharts-block dashboard-tile-shadow"
              data-testid="chart-region"
              role="region"
            >
              <div
                data-testid="chart"
              />
            </div>
            
          </div>
        </div>
        
      </div>
      <div
        class="row"
        data-testid="row"
      >
        
        <div
          class="my-2 col-lg-4"
          data-testid="col"
        >
          <div
            data-testid="block-renderer"
          >
            
            <div
              aria-label="Patient by # of Assessment Completed"
              class="highcharts-block dashboard-tile-shadow"
              data-testid="chart-region"
              role="region"
            >
              <div
                data-testid="chart"
              />
            </div>
            
          </div>
        </div>
        <div
          class="my-2 col-lg-8"
          data-testid="col"
        >
          <div
            data-testid="block-renderer"
          >
            
            <div
              aria-label="Addiction Medicine Program – Substance Use"
              class="callout-block w-100 d-flex align-items-center justify-content-center dashboard-tile-height"
              data-testid="callout"
              role="region"
            >
              <div
                class="px-4"
              >
                <div
                  class="text-mulberry-dark font-size-md pb-2"
                >
                  Addiction Medicine Program – Substance Use
                </div>
                <div
                  class="text-grey-darkest font-size-xs"
                >
                  This concurrent (co-occurring) program allows individuals time to fully explore the intersection of their mental health and substance use issues while learning skills they need for lasting recovery. The connections between mental health and substance use can be complex, and both issues deserve and require evidence-based therapies to support recovery and wellness.
                </div>
              </div>
            </div>
            
          </div>
        </div>
        
      </div>
      <div
        class="row"
        data-testid="row"
      >
        
        <div
          class="my-2 col-lg-8"
          data-testid="col"
        >
          <div
            data-testid="block-renderer"
          >
            
            <div
              aria-label="Patient by # of Assessment Completed"
              class="highcharts-block dashboard-tile-shadow"
              data-testid="chart-region"
              role="region"
            >
              <div
                data-testid="chart"
              />
            </div>
            
          </div>
        </div>
        <div
          class="my-2 col-lg-4"
          data-testid="col"
        >
          <div
            data-testid="block-renderer"
          >
            
            <div
              aria-label="Recovery"
              class="highcharts-block dashboard-tile-shadow"
              data-testid="chart-region"
              role="region"
            >
              <div
                data-testid="chart"
              />
            </div>
            
          </div>
        </div>
        
      </div>
      <div
        class="row"
        data-testid="row"
      >
        
        <div
          class="my-2 col-lg-12"
          data-testid="col"
        >
          <div
            data-testid="block-renderer"
          >
            
            <div
              aria-label="Symptom Severity Flow"
              class="highcharts-block dashboard-tile-shadow"
              data-testid="chart-region"
              role="region"
            >
              <div
                data-testid="chart"
              />
            </div>
            
          </div>
        </div>
        
      </div>
      
    </div>
    
  </div>
</div>
`;
